db/migrate/20170923045153_change_user_address_column_names.rb

```diff
@@ -0,0 +1,22 @@
+class ChangeUserAddressColumnNames < ActiveRecord::Migration[5.0] 
+  def up
+    incompatible_records = execute("SELECT id, name, email, state FROM users WHERE LENGTH(state) > 2;")
+    raise "Migration aborted beacuse legacy data contains users with states that are more than two letters: #{incompatible_records.to_a}" unless incompatible_records.to_a.empty?
```

Comment:
this seems weird to just bomb out, @markottaviani can we get a dump or something to clean these up before this runs?

Also why are you doing raw sql here?  You can do some fancy stuff instead

`User.where('LENGTH(state) > 2').exists?`

Reply:
> this seems weird to just bomb out

I don't know how you guys typically handle data migrations, so this is only here to make it abundantly clear that this could theoretically be dangerous (destroy production data). I thought it was safer for the migration to fail under the described condition. If the data is checked before migrating, then it's not technically necessary.

> can we get a dump or something to clean these up before this runs

My intention was that this query would provide enough information to be able to identify any users whose state violates this constraint (if there are any). 

> Also why are you doing raw sql here? You can do some fancy stuff instead

The short answer is because I knew how to get the result I wanted faster with SQL than with ActiveRecord. 

Do you prefer "fancy stuff"? Introducing models into migrations has been something that has been non-trivial in my experience, as the `User` model is not necessarily defined in the way you'd expect when this migration runs. The guide touches on this - see [here](http://guides.rubyonrails.org/v3.2/migrations.html#using-models-in-your-migrations)

>When creating or updating data in a migration it is often tempting to use one of your models. After all, they exist to provide easy access to the underlying data. This can be done, but some caution should be observed.

Some people re-declare the models in the migrations to address this, which is another complication. In this specific case I don't imagine it matters one way or another. I've generally found it's easier to default to avoiding models where they're not necessary, as in my view there's nothing wrong or undesirable about SQL. 

All that to say, if the SQL is simple is there a downside to it/is it less desirable for you?

Reply:
SQL should be fine.  Fancy stuff is good too. We run a data migration gem or would make a rake task to do something like this since it's a 1 off and not really important to be part of the db schema definition. In those you'd have access to the models.

Careful with that doc link you posted, it's for Rails 3.2, we're on 5.

Reply:
I'm not clear on the outcome you'd like based on what you said - are you wanting the `select` removed from this file, as it'll be handled outside the migrations? It's not part of the schema definition as presently implemented. 

> Careful with that doc link you posted, it's for Rails 3.2, we're on 5.

Yeah, that was the first result from a google search. As a resource for that particular concern it's still valid, as far as I know nothing has changed in that department. If you're looking for more info on the issue more recent sources may be more useful.

Reply:
I should also note that I apologize if I’m talking down to you, no clue what your experience level is, looking forward to getting to know that at our meeting soon!

I think for this we want to separate the renaming of the column and the fixing of the data. I would say that the steps here are

1. Prevent anything bigger than 2 characters into state with a rails model validation.
2. See if any data needs fixing, if it does fix it with a rake task or direct sql.
3. Put constraints on the column.
4. Rename column.

3 and 4 can be the same migration.

What happens after the column rename to the alias_attribute now that you are aliasing a column that no longer exists?

We could just do without that as the migration would be quick and we haven’t been going for 0 downtime migrations. If we were you would do something like https://github.com/ankane/strong_migrations/blob/master/README.md#renaming-or-changing-the-type-of-a-column.

Reply:
No stress! I appreciate you being as thorough as possible, as it will help me with on-boarding. I didn't assume you were talking up or down. I hope I didn't come across as offended, I'm making an effort to explain my reasoning behind everything to try and short circuit any confusion.

> Prevent anything bigger than 2 characters into state with a rails model validation.
@markottaviani referenced that this is already in place on the `User` model

> See if any data needs fixing, if it does fix it with a rake task or direct sql.
Sounds like that's all good as well

> Put constraints on the column.
Rename column.

Sounds good

> What happens before the column rename to the alias_attribute when it is referencing a column that doesn’t exist??

You mean if the `User` model is deployed and the server is started before the database is migrated? I imagine almost all workflows would break, as you'd end up the same as any other missing attribute, e.g. `NoMethodError: undefined method 'region=' for #<User:0x00000008043cf8>` for the setter and the same for the getter.

Reply:
Our capistrano setup migrates the database changes first